File: SumOfNodesInBSTRange/SumOfNodesInBSTRange.cpp
/*
https://www.geeksforgeeks.org/problems/range-sum-of-bst/1
*/

/*
class Node {
  public:
    int data;
    Node* left;
    Node* right;

    // Constructor to initialize a new node
    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};
*/

class Solution {
  public:
  
    void dfs(Node *root, int l, int r, int &res){
        if(root==nullptr) return;
        if(root->data>=l&&root->data<=r){
            res+=root->data;
        }
        dfs(root->left, l, r, res);
        dfs(root->right, l, r, res);
        return;
    }
  
    void preOrderTraversal(Node *root, int l, int r, int &res){
        dfs(root, l, r, res);
    }
  
    int nodeSum(Node* root, int l, int r) {
        int res=0;
        preOrderTraversal(root, l, r, res);
        return res;
    }
};

File: SumOfNodesInBSTRange/SumOfNodesInBSTRange.java
/*
https://www.geeksforgeeks.org/problems/range-sum-of-bst/1
*/

/*
class Node {
    int data;
    Node left, right;
    Node(int val) {
        data = val;
        left = right = null;
    }
}
*/

class Solution {
    
    private int resultantRangeSum;
    
    public int getResultantRangeSum(){
        return this.resultantRangeSum;
    }
    
    public void setResultantRangeSum(int resultantRangeSum){
        this.resultantRangeSum=resultantRangeSum;
    }
    
    public void depthFirstSearch_forTrees(Node root, int l, int r){
        if(root==null) return;
        if(root.data>=l&&root.data<=r){
            int currentRangeSum=getResultantRangeSum();
            currentRangeSum+=root.data;
            setResultantRangeSum(currentRangeSum);
        }
        depthFirstSearch_forTrees(root.left, l, r);
        depthFirstSearch_forTrees(root.right, l, r);
        return;
    }
    
    public void preOrderTraversal(Node root, int l, int r){
        depthFirstSearch_forTrees(root, l, r);
    }
    
    public int nodeSum(Node root, int l, int r) {
        preOrderTraversal(root, l, r);
        return getResultantRangeSum();
    }
}
